### DISCUSSION TITLE
Multi threading when used nat-nat ipsec based on swanctl

### DISCUSSION THREAD

**Discussion by alex-airspan** [strongswan/strongswan#1063] [Q&A]
Hi

I use strongSwan swanctl version 5.7.2 in nat-nat with virtual ipsec0 interface ( according to tests in strongswan site) 

Tunnel opened and traffic forwarded. but MAX traffic that I reach ~ 100Mbps. In this state I see one of threads reach to 100% CPU and second thread os second CPU reach to ~30% CPU 

My question - if possible improve the performance with divide the first thread to two or more cores

Thanks,
Alex

**Comment by tobiasbrunner** [comment]
> with virtual ipsec0 interface ( according to tests in strongswan site)

What exactly do you refer to here? There are several approaches for route-based VPNs that use different types of interfaces (see [here](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html)). In particular, if you refer to _kernel-libipsec_ then its limitations are clearly documented [here](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html). For other types of interfaces it really depends on a lot of things (CPU, NIC, kernel version, policy/route settings etc.).

**Comment by alex-airspan** [comment]
Hi Tobias,

Yes, We refer kernel-libipsec and you correct - limitation documented. But it really ~120Mbps only? 

An additional - our company ( Airspan Networks) want use strongswan for IPv6 with any special configuration of k0s cluster. 
We tried some configuration based on route-based VPNs and current we  found only this.

Maybe possible give us dedicated support? If need - we will pay about it. 

Thanks,
Alex

**Reply by tobiasbrunner** [reply]
> But it really ~120Mbps only?

No idea, but sounds reasonable enough depending on the CPU etc. It's simply two threads that process in- and outbound packets and there is not much optimization going on in regards to locking/queuing etc. As documented, it's just not intended for high-performance uses.

So if you want to use route-based VPNs, try one of the kernel-based solutions (or use policy-based VPNs).

**Comment by alex-airspan** [comment]
++

Hi Tobias,

Thanks for answer

What you mean - kernel-based solutions?

What about give use dedicated support for found the E2E solution?

Our limitation – we Centos7 with below kernel

***@***.*** ~]# uname -a
Linux localhost.localdomain 3.10.0-1160.62.1.rt56.1203.1.at2200.el7.x86_64 #1 SMP PREEMPT RT Thu Apr 28 14:31:21 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
***@***.*** ~]#


Thanks,
Alex

**Reply by tobiasbrunner** [reply]
> What you mean - kernel-based solutions?

I linked the page that describes the different approaches above.

> What about give use dedicated support for found the E2E solution?

No idea what you mean, but see [here](https://docs.strongswan.org/docs/5.9/support/commercial.html).

> Our limitation – we Centos7 with below kernel

So you shoot yourself in the foot and then come here to complain.

**Comment by alex-airspan** [comment]
What your recommendation for  Realtime kernel + OS for embedded devices  + IPv6 + Strongswan (route base) kernel base ?

Thanks

**Reply by tobiasbrunner** [reply]
No experience with RT (or how it affects IPsec, policy- or route-based) but for route-based VPNs you probably want to use a kernel that supports XFRM interfaces (i.e. >= 4.19).
